### PR TITLE
deepautoqsar changes

### PR DESCRIPTION
**Pull request by leelaprabhu**
Changes needed for DeepAutoQSAR

### CODE REVIEW

File: deepchem/models/tensorgraph/layers.py [hunk 1]
```diff
@@ -983,6 +989,22 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
       self.out_tensor = out_tensor
     return out_tensor
 
+def geqk(x,k):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:992]: what?  can you leave a comment or rename something more readable?

File: deepchem/models/tensorgraph/layers.py [hunk 2]
```diff
@@ -983,6 +989,22 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
...
+  dim = int(x.shape[1])
+  x_T = tf.transpose(x)
+  k_max = tf.nn.top_k(x_T, k=k, sorted=True)
+  flat = tf.reshape(k_max[0], [-1, (k*dim)])
+  return flat
+
+def lessk(x,k):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:999]: what is this doing?

File: deepchem/models/tensorgraph/layers.py [hunk 3]
```diff
@@ -983,6 +989,22 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
...
+  top = tf.reduce_max(x, 0, keep_dims=True)
+  repl = tf.concat(axis=1,values=[top]*k)
+  return repl
+
+def reduce_max_k(x,k):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:1004]: what is this doing?

File: deepchem/models/tensorgraph/layers.py [hunk 4]
```diff
@@ -1486,18 +1509,11 @@ def get_cells(self):
...
-
-  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
-    inputs = self._get_input_tensors(in_layers)
-    parent_tensor = inputs[0]
-    keep_prob = 1.0 - self.dropout_prob * kwargs['training']
-    out_tensor = tf.nn.dropout(parent_tensor, keep_prob)
-    if set_tensors:
-      self.out_tensor = out_tensor
-    return out_tensor
+  def create_tensor(self):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:1512]: why these changes?

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 5]
```diff
@@ -323,7 +323,7 @@ def build(self):
       for node in order:
         with tf.name_scope(node):
           node_layer = self.layers[node]
-          node_layer.create_tensor(training=self._training_placeholder)
+          node_layer.create_tensor()
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tensor_graph.py:326]: what was wrong with the original dropout implementation?

File: deepchem/models/tensorgraph/layers.py (hunk 1 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:992]: Is this the best place for this?

File: deepchem/models/tensorgraph/layers.py [hunk 6]
```diff
@@ -507,20 +507,26 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
 
 class Concat(Layer):
 
-  def __init__(self, axis=1, **kwargs):
+  def __init__(self, axis=1, normalize=False, svd=False, **kwargs):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:510]: maybe variable should be l2normalize not just normalize

File: deepchem/data/datasets.py [hunk 7]
```diff
@@ -323,8 +323,7 @@ def iterate(dataset, batch_size, deterministic, pad_batches):
         sample_perm = np.arange(n_samples)
       if batch_size is None:
         batch_size = n_samples
-      interval_points = np.linspace(
-          0, n_samples, np.ceil(float(n_samples) / batch_size) + 1, dtype=int)
+      interval_points = np.asarray([min(i*batch_size, n_samples) for i in range(1+int(np.ceil(n_samples/batch_size)))])
```
Comment (lilleswing) [inline deepchem/data/datasets.py:326]: can you add a test for this?

File: deepchem/models/tensorgraph/tensor_graph.py (hunk 5 above)
Comment (leelaprabhu) [reply deepchem/models/tensorgraph/tensor_graph.py:326]: Nothing really, you can't dynamically change the keep_prob/dropout_prob- but I don't think that's important. 
I'll change my DeepAutoQSAR code to match the DeepChem dropout layer and then I can remove my version.

File: deepchem/models/tensorgraph/layers.py (hunk 1 above)
Comment (leelaprabhu) [reply deepchem/models/tensorgraph/layers.py:992]: Where should I add this? I think I should add it to GraphGather, would that be ok?
